In [1]:
HTML(read(open("style.css"), String))

HTML{String}("<link href='http://fonts.googleapis.com/css?family=Alegreya+Sans:100,300,400,500,700,800,900,100italic,300italic,400italic,500italic,700italic,800italic,900italic' rel='stylesheet' type='text/css'>\r\n<link href='http://fonts.googleapis.com/css?family=Arvo:400,700,400italic' rel='stylesheet' type='text/css'>\r\n<link href='http://fonts.googleapis.com/css?family=PT+Mono' rel='stylesheet' type='text/css'>\r\n<link href='http://fonts.googleapis.com/css?family=Shadows+Into+Light' rel='stylesheet' type='text/css'>\r\n<link href='http://fonts.googleapis.com/css?family=Philosopher:400,700,400italic,700italic' rel='stylesheet' type='text/css'>\r\n\r\n<style>\r\n\r\n@font-face {\r\n    font-family: \"Computer Modern\";\r\n    src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');\r\n}\r\n\r\n.container { width: 100% }\r\n\r\n/* Formatting for header cells */\r\n.text_cell_render h1 {\r\n    font-family: 'Philosopher', sans-serif;\r\n    font-weight: 400;\r\n    font-size: 2.2em;\r\n    line-height: 100%;\r\n    color: rgb(0, 80, 120);\r\n    margin-bottom: 0.1em;\r\n    margin-top: 0.1em;\r\n    display: block;\r\n}\t\r\n.text_cell_render h2 {\r\n    font-family: 'Philosopher', serif;\r\n    font-weight: 400;\r\n    font-size: 1.9em;\r\n    line-height: 100%;\r\n    color: rgb(200,100,0);\r\n    margin-bottom: 0.1em;\r\n    margin-top: 0.1em;\r\n    display: block;\r\n}\t\r\n\r\n.text_cell_render h3 {\r\n    font-family: 'Philosopher', serif;\r\n    margin-top:12px;\r\n    margin-bottom: 3px;\r\n    font-style: italic;\r\n    color: rgb(94,127,192);\r\n}\r\n\r\n.text_cell_render h4 {\r\n    font-family: 'Philosopher', serif;\r\n}\r\n\r\n.text_cell_render h5 {\r\n    font-family: 'Alegreya Sans', sans-serif;\r\n    font-weight: 300;\r\n    font-size: 16pt;\r\n    color: grey;\r\n    font-style: italic;\r\n    margin-bottom: .1em;\r\n    margin-top: 0.1em;\r\n    display: block;\r\n}\r\n\r\n.text_cell_render h6 {\r\n    font-family: 'PT Mono', sans-serif;\r\n    font-weight: 300;\r\n    font-size: 10pt;\r\n    color: grey;\r\n    margin-bottom: 1px;\r\n    margin-top: 1px;\r\n}\r\n\r\n.text_cell_render em {\r\n    font-family: 'Philosopher', sans-serif;\r\n    color:        blue;\r\n    background-color: rgb(255,220,180);\r\n    font-size:    110%;\r\n    margin-left:   2px;\r\n    margin-right:  2px;\r\n    font-weight:   100;\r\n}\r\n\r\n.text_cell_render b {\r\n    color:            rgb(255,195,195);\r\n    background-color: rgb(0,0,0);\r\n    font-size:    110%;\r\n    margin-left:   2px;\r\n    margin-right:  2px;\r\n    font-weight:   650;\r\n}\r\n\r\n.text_cell_render tt {\r\n    font-size:    120%;\r\n    margin-left:   2px;\r\n    margin-right:  2px;\r\n    font-weight:   150;\r\n}\r\n\r\n.Codemirror {\r\n    font-family: \"PT Mono\";\r\n    font-size: 100%;\r\n}\r\n\r\n</style>\r\n\r\n")

## Tests for Minimax.ipynb

This notebook contains all tests from the `Minimax.ipynb` jupyter notebook.

In [2]:
# Pkg.add("NBInclude")
using NBInclude
# Pkg.add("Chess")
using Chess
using Test

In [3]:
@nbinclude("Minimax.ipynb")

minimax (generic function with 2 methods)

## Minimax incremental without Memoization

In [4]:
# Henry Buckle vs NN, London, 1840
b1 = fromfen("r2qkb1r/pp2nppp/3p4/2pNN1B1/2BnP3/3P4/PPP2PPP/R2bK2R w KQkq - 1 0")
# 1. Nf6+ gxf6 2. Bxf7# 

# Louis Paulsen vs Blachy, New York, 1857
b2 = fromfen("1rb4r/pkPp3p/1b1P3n/1Q6/N3Pp2/8/P1P3PP/7K w - - 1 0")
# 1. Qd5+ Ka6 2. cxb8=N# 

# Paul Morphy vs Duke Isouard, Paris, 1858
b3 = fromfen("4kb1r/p2n1ppp/4q3/4p1B1/4P3/1Q6/PPP2PPP/2KR4 w k - 1 0")
# 1. Qb8+ Nxb8 2. Rd8# 

# Johannes Zukertort vs Adolf Anderssen, Breslau, 1865
b4 = fromfen("r1b2k1r/ppp1bppp/8/1B1Q4/5q2/2P5/PPP2PPP/R3R1K1 w - - 1 0")
# 1. Qd8+ Bxd8 2. Re8# 

# Gustav Neumann vs Carl Mayet, Berlin, 1866
b5 = fromfen("5rkr/pp2Rp2/1b1p1Pb1/3P2Q1/2n3P1/2p5/P4P2/4R1K1 w - - 1 0")
# 1. Qxg6+ fxg6 2. Rg7# 

@testset "minimax" begin
    display(b1)
    @test minimax_noMem(b1, evaluate_position(b1), 4)[2] == movefromstring("d5f6")
    display(b2) 
    @test minimax_noMem(b2, evaluate_position(b1), 4)[2] == movefromstring("b5d5")
    #display(b3)
    #@test minimax_noMem(b3, evaluate_position(b1), 4)[2] == movefromstring("b3b8")
    #display(b4)
    #@test minimax_noMem(b4, evaluate_position(b1), 4)[2] == movefromstring("d5d8")
    #display(b5)
    #@test minimax_noMem(b5, evaluate_position(b1), 4)[2] == movefromstring("g5g6")
end

Board (r2qkb1r/pp2nppp/3p4/2pNN1B1/2BnP3/3P4/PPP2PPP/R2bK2R w KQkq -):
 r  -  -  q  k  b  -  r 
 p  p  -  -  n  p  p  p 
 -  -  -  p  -  -  -  - 
 -  -  p  N  N  -  B  - 
 -  -  B  n  P  -  -  - 
 -  -  -  P  -  -  -  - 
 P  P  P  -  -  P  P  P 
 R  -  -  b  K  -  -  R

Board (1rb4r/pkPp3p/1b1P3n/1Q6/N3Pp2/8/P1P3PP/7K w - -):
 -  r  b  -  -  -  -  r 
 p  k  P  p  -  -  -  p 
 -  b  -  P  -  -  -  n 
 -  Q  -  -  -  -  -  - 
 N  -  -  -  P  p  -  - 
 -  -  -  -  -  -  -  - 
 P  -  P  -  -  -  P  P 
 -  -  -  -  -  -  -  K

Test Summary: | Pass  Total  Time
minimax       |    2      2  5.4s


Test.DefaultTestSet("minimax", Any[], 2, false, false, true, 1.681649082358e9, 1.681649087735e9)

## Minimax with Memoization

In [5]:
# Henry Buckle vs NN, London, 1840
b1 = fromfen("r2qkb1r/pp2nppp/3p4/2pNN1B1/2BnP3/3P4/PPP2PPP/R2bK2R w KQkq - 1 0")
# 1. Nf6+ gxf6 2. Bxf7# 
a1 = AdvBoard(b1)

# Louis Paulsen vs Blachy, New York, 1857
b2 = fromfen("1rb4r/pkPp3p/1b1P3n/1Q6/N3Pp2/8/P1P3PP/7K w - - 1 0")
# 1. Qd5+ Ka6 2. cxb8=N# 
a2 = AdvBoard(b2)

# Paul Morphy vs Duke Isouard, Paris, 1858
b3 = fromfen("4kb1r/p2n1ppp/4q3/4p1B1/4P3/1Q6/PPP2PPP/2KR4 w k - 1 0")
# 1. Qb8+ Nxb8 2. Rd8# 
a3 = AdvBoard(b3)

# Johannes Zukertort vs Adolf Anderssen, Breslau, 1865
b4 = fromfen("r1b2k1r/ppp1bppp/8/1B1Q4/5q2/2P5/PPP2PPP/R3R1K1 w - - 1 0")
# 1. Qd8+ Bxd8 2. Re8# 
a4 = AdvBoard(b4)

# Gustav Neumann vs Carl Mayet, Berlin, 1866
b5 = fromfen("5rkr/pp2Rp2/1b1p1Pb1/3P2Q1/2n3P1/2p5/P4P2/4R1K1 w - - 1 0")
# 1. Qxg6+ fxg6 2. Rg7# 
a5 = AdvBoard(b5)

@testset "minimax" begin
    display(b1)
    @test minimax(a1, 4)[2] == movefromstring("d5f6")
    display(b2)
    @test minimax(a2, 4)[2] == movefromstring("b5d5")
    display(b3)
    @test minimax(a3, 4)[2] == movefromstring("b3b8")
    display(b4)
    @test minimax(a4, 4)[2] == movefromstring("d5d8")
    display(b5)
    @test minimax(a5, 4)[2] == movefromstring("g5g6")
end

Board (r2qkb1r/pp2nppp/3p4/2pNN1B1/2BnP3/3P4/PPP2PPP/R2bK2R w KQkq -):
 r  -  -  q  k  b  -  r 
 p  p  -  -  n  p  p  p 
 -  -  -  p  -  -  -  - 
 -  -  p  N  N  -  B  - 
 -  -  B  n  P  -  -  - 
 -  -  -  P  -  -  -  - 
 P  P  P  -  -  P  P  P 
 R  -  -  b  K  -  -  R

Board (1rb4r/pkPp3p/1b1P3n/1Q6/N3Pp2/8/P1P3PP/7K w - -):
 -  r  b  -  -  -  -  r 
 p  k  P  p  -  -  -  p 
 -  b  -  P  -  -  -  n 
 -  Q  -  -  -  -  -  - 
 N  -  -  -  P  p  -  - 
 -  -  -  -  -  -  -  - 
 P  -  P  -  -  -  P  P 
 -  -  -  -  -  -  -  K

Board (4kb1r/p2n1ppp/4q3/4p1B1/4P3/1Q6/PPP2PPP/2KR4 w k -):
 -  -  -  -  k  b  -  r 
 p  -  -  n  -  p  p  p 
 -  -  -  -  q  -  -  - 
 -  -  -  -  p  -  B  - 
 -  -  -  -  P  -  -  - 
 -  Q  -  -  -  -  -  - 
 P  P  P  -  -  P  P  P 
 -  -  K  R  -  -  -  -

Board (r1b2k1r/ppp1bppp/8/1B1Q4/5q2/2P5/PPP2PPP/R3R1K1 w - -):
 r  -  b  -  -  k  -  r 
 p  p  p  -  b  p  p  p 
 -  -  -  -  -  -  -  - 
 -  B  -  Q  -  -  -  - 
 -  -  -  -  -  q  -  - 
 -  -  P  -  -  -  -  - 
 P  P  P  -  -  P  P  P 
 R  -  -  -  R  -  K  -

Board (5rkr/pp2Rp2/1b1p1Pb1/3P2Q1/2n3P1/2p5/P4P2/4R1K1 w - -):
 -  -  -  -  -  r  k  r 
 p  p  -  -  R  p  -  - 
 -  b  -  p  -  P  b  - 
 -  -  -  P  -  -  Q  - 
 -  -  n  -  -  -  P  - 
 -  -  p  -  -  -  -  - 
 P  -  -  -  -  P  -  - 
 -  -  -  -  R  -  K  -

Test Summary: | Pass  Total   Time
minimax       |    5      5  11.3s


Test.DefaultTestSet("minimax", Any[], 5, false, false, true, 1.681649088164e9, 1.68164909942e9)

### Results

In [6]:
# Ding Liren vs Ian Nepomniachtchi, 2023
b = fromfen("r1bq1rk1/ppp2ppp/2np4/8/2PPPp2/2P2N2/P1Q1BPPP/R3K2R w KQ - 0 11")
a = AdvBoard(b)

display(b)
@time minimax(a, 5)

Board (r1bq1rk1/ppp2ppp/2np4/8/2PPPp2/2P2N2/P1Q1BPPP/R3K2R w KQ -):
 r  -  b  q  -  r  k  - 
 p  p  p  -  -  p  p  p 
 -  -  n  p  -  -  -  - 
 -  -  -  -  -  -  -  - 
 -  -  P  P  P  p  -  - 
 -  -  P  -  -  N  -  - 
 P  -  Q  -  B  P  P  P 
 R  -  -  -  K  -  -  R

 28.521207 seconds (219.76 M allocations: 15.359 GiB, 9.77% gc time)


(110, Move(a1b1))

1. $d = 1$: (20, Move(e1g1))   0.000355 seconds (959 allocations: 80.172 KiB)
1. $d = 2$: (-25, Move(d4d5))  0.004071 seconds (18.84 k allocations: 1.825 MiB)
1. $d = 3$: (90, Move(a1b1))   0.105581 seconds (548.58 k allocations: 46.607 MiB)
1. $d = 4$: (-110, Move(d4d5)) 4.531544 seconds (12.10 M allocations: 920.138 MiB, 21.99% gc time)
1. $d = 5$: (110, Move(a1b1))  76.235485 seconds (219.80 M allocations: 15.360 GiB, 20.58% gc time)